src/dune_rules/simple_rules.ml

```diff
@@ -59,15 +59,27 @@ let interpret_and_add_locks ~expander locks action =
   | locks -> Action_builder.map action ~f:(Action.Full.add_locks locks)
 ;;
 
+let mark_as_unsafe_for_shared_cache ~cache_user_rules action =
+  let open Action_builder.O in
+  let+ action = action in
+  Action.Full.add_can_go_in_shared_cache cache_user_rules action
+;;
+
 let add_user_rule
   sctx
   ~dir
   ~(rule : Rule_conf.t)
-  ~(action : _ Action_builder.With_targets.t)
+  ~(action : Action.Full.t Action_builder.With_targets.t)
   ~expander
   =
+  let open Memo.O in
+  let* config = Workspace.workspace_config () in
```

Comment:
Note that once the cache setting is set, it can no longer be modified in watch mode. There's a few options like that in the workspace that are incompatible with the watch mode.

So this is not better than just using a global var. It's slightly confusing in fact as it pretends to be something that it's not (a value that can change in watch mode)